### DISCUSSION TITLE
IKE authentication credentials unacceptable on Windows

### DISCUSSION THREAD

**Discussion by SaiPremaTech** [strongswan/strongswan#1351] [Q&A]
**System (please complete the following information):**
 - OS: CentOS Linux release 7.9.2009 (Core)
 - Kernel version (if applicable): [e.g. 5.10]
 - strongSwan version(s): Linux strongSwan U5.7.2/K3.10.0-1160.76.1.el7.x86_64
 - Tested/confirmed with the latest version: no

**Describe the bug**
I am trying to get the Windows built-in IKE2 VPN client connected with our StrongSwan server. I have tried EAP configuration on both windows 10 and 7. I have been following this link - https://docs.strongswan.org/docs/5.9/interop/windowsEapConf.html. I keep seeing this error - 13801 IKE authentication credentials are unacceptable on both Windows 10 and 7 clients. Here is the server log and IPSec configuration. I have used self signed root CA and server certificates. Please let me know if I am missing anything. Any directions and help is appreciated.

Thanks and Regards,
Ashok

[windows-ikev2-vpn-client-connectivity-issue.txt](https://github.com/strongswan/strongswan/files/9814174/windows-ikev2-vpn-client-connectivity-issue.txt)

**Comment by Thermi** [comment]
eap_identity=amdemo_test1
That needs to be %any because it asks for the client eap identity.

**Reply by sendel** [reply]
> Oct 18 15:20:34 ip-10-101-1-6 charon: 07[ENC] generating IKE_AUTH response 1 [ IDr CERT AUTH EAP/REQ/MSCHAPV2 ]
> Oct 18 15:20:34 ip-10-101-1-6 charon: 07[ENC] splitting IKE message (1564 bytes) into 2 fragments
> Oct 18 15:20:34 ip-10-101-1-6 charon: 07[ENC] generating IKE_AUTH response 1 [ EF(1/2) ]
> Oct 18 15:20:34 ip-10-101-1-6 charon: 07[ENC] generating IKE_AUTH response 1 [ EF(2/2) ]
> Oct 18 15:20:34 ip-10-101-1-6 charon: 07[NET] sending packet: from 10.101.1.6[4500] to 50.2.0.2[4500] (1248 bytes)
> Oct 18 15:20:34 ip-10-101-1-6 charon: 07[NET] sending packet: from 10.101.1.6[4500] to 50.2.0.2[4500] (384 bytes)

Im got this too. Its trouble with certificate. Cert for Windows must be in .p12 with secretkey
Example if certs processed for OpenWrt 23+ by strongswan-gencerts:

`    #openssl pkcs12 -export -inkey ./private/srv_crts.key -in ./x509/srv_crts.crt -name "client" -certfile ./x509ca/violonix.crt -caname "Server VPN Root CA" -out serverCert.p12`

**Comment by SaiPremaTech** [comment]
Thanks Thermi. I have tried setting eap_identity to %any. I still see the same error on the client.

**Comment by Thermi** [comment]
Please provide updated logs with the new setting. There might be more wrong.

**Comment by SaiPremaTech** [comment]
Thank you so much Thermi. Here is the file containing the latest log with the change that you suggested.

[windows-ikev2-vpn-client-connectivity-issue-log-2.txt](https://github.com/strongswan/strongswan/files/9815676/windows-ikev2-vpn-client-connectivity-issue-log-2.txt)

**Comment by Thermi** [comment]
the log ends too soon. Also considering you are using eap-radius we will also need to consider what the RADIUS server itself does and its reasoning for the reply, as soon as we get to that stage. Please provide longer logs. Also maybe disable the syslog logger in strongSwan, you have duplicate log lines right now.

**Comment by SaiPremaTech** [comment]
I don't see any other log messages other than 'deleting the half open IKE_SA' as you can see in the attached file. This is for windows 10 client attempt. Also, there is no attempt made with Radius server as we were seeing earlier when I set eap_identity=amdemo_test1.

[windows-ikev2-vpn-client-connectivity-issue-log-3.txt](https://github.com/strongswan/strongswan/files/9815881/windows-ikev2-vpn-client-connectivity-issue-log-3.txt)

The following log is for the attempt made from windows 7 client. it's also throwing the same error on the client side (i.e. 13801 - IKE authentication credentials are unacceptable)
[windows-7-ikev2-vpn-client-connectivity-issue-log-4.txt](https://github.com/strongswan/strongswan/files/9815908/windows-7-ikev2-vpn-client-connectivity-issue-log-4.txt)

**Reply by tobiasbrunner** [reply]
Your Windows 10 client doesn't seem to have the necessary CA certificate installed (at least it sends no certificate request for it, only 69 for other CAs). The Windows 7 client does, but there could be an IP fragmentation issue (the IKE_AUTH response is just a bit larger than 1500 bytes and Windows 7 does not support IKEv2 fragmentation, which Windows 10 does).

**Comment by Brink2Three** [comment]
I realize it's a few months later but I have an addition to add to this: (I can move this to it's own post if needed) 
While building out a similar configuration, I'm getting error 13801 on windows clients as well. Specifically it seems to be related to the IKE Name EKU check run by windows when checking the certificate. 

In registry, if I go to `Computer\HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\RasMan\Parameters` and set `DisableIKENameEkuCheck` to 1, the connection works. 

I tried generating the certificates using the method described in this [Server Fault post](https://serverfault.com/questions/536092/strongswan-ikev2-windows-7-agile-vpn-what-is-causing-error-13801), but as it's 10 years old, I suspect it is outdated, and when imported into windows, I get an error that Windows can't identify the certificate. 

The flags are set properly when generating the key... (serverAuth) Any ideas?

**Reply by tobiasbrunner** [reply]
> and when imported into windows, I get an error that Windows can't identify the certificate

During the import? What do you mean exactly with "can't identity"? And what certificate did you try importing?

> but as it's 10 years old, I suspect it is outdated

There is a redundant step that extracts the public key (`pki --issue` can directly work with private keys) and an unnecessary step that converts the key to DER format (in particular because the key was generated in PEM format explicitly, strongSwan can import DER or PEM anyway), but other than that it looks OK.

> The flags are set properly when generating the key...

Hm, maybe that's a typo, but the `serverAuth` flag has to be passed when issuing the certificate, not when generating the key.

Anyway, please post the output of `pki --print` for the certificates of your whole PKI and more details on what you tried to import on Windows.

**Reply by Brink2Three** [reply]
> During the import? What do you mean exactly with "can't identity"? And what certificate did you try importing?

The Root CA, after importing into windows. Here's a screenshot. If I were to follow his steps exactly and keep names the same, the file I imported into windows would be `vpn.example.com.crt.pem`.
![image](https://user-images.githubusercontent.com/47770944/222469416-976947eb-0c1a-4d12-b804-d3e1860b9d54.png)

I did try a different tutorial from [Digital Ocean](https://www.digitalocean.com/community/tutorials/how-to-set-up-an-ikev2-vpn-server-with-strongswan-on-ubuntu-20-04), and while the one generated by their steps (2 and 3) is correctly identified in windows, I still get error 13801 until I disable EKU checks. 

I should note that windows identifies "Digital ocean" cert Key Usage as an `<All>` cert, while the Server Fault cert ONLY has the EKU tag for `serverAuth`. 

> please post the output of pki --print for the certificates of your whole PKI

Here are the PKI print outputs of each cert that is imported into windows:
Serverfault steps: 
```
pki --print --in serverfault.crt.pem
  subject:  "C=US, O=Example Company, CN=vpn.vpnexample.net"
  issuer:   "C=US, O=Example Company, CN=Example VPN CA"
  validity:  not before Mar 02 15:28:48 2023, ok
             not after  Mar 01 15:28:48 2026, ok (expires in 1094 days)
  serial:    36:7e:46:3a:3e:e8:b5:52
  altNames:  vpn.vpnexample.net
  flags:     serverAuth
  authkeyId: d7:b1:a8:0c:ea:ec:34:73:41:57:24:8b:67:f7:75:1e:f3:0d:9e:77
  subjkeyId: ed:1c:9e:62:4d:cb:4e:0d:46:90:f7:17:54:54:37:c2:1a:20:78:85
  pubkey:    RSA 4096 bits
  keyid:     d2:df:90:f1:1a:8b:2f:54:a3:f2:f7:c6:c2:b4:ff:8a:44:94:31:22
  subjkey:   ed:1c:9e:62:4d:cb:4e:0d:46:90:f7:17:54:54:37:c2:1a:20:78:85
```
Digital Ocean steps:
```
pki --print --in ca-cert.pem
  subject:  "CN=vpn.vpnexample.net"
  issuer:   "CN=VPN root CA"
  validity:  not before Mar 01 23:16:26 2023, ok
             not after  Feb 26 23:16:26 2033, ok (expires in 3649 days)
  serial:    6e:bc:8b:c6:8f:15:33:1c
  flags:     CA CRLSign self-signed
  subjkeyId: 2a:14:bc:98:86:47:ec:0b:7d:67:b0:ff:a2:3d:a5:e5:de:52:ba:c6
  pubkey:    RSA 4096 bits
  keyid:     ab:b7:8e:32:5a:b8:24:12:9a:77:76:02:80:ec:5e:6f:66:08:08:b2
  subjkey:   2a:14:bc:98:86:47:ec:0b:7d:67:b0:ff:a2:3d:a5:e5:de:52:ba:c6
```
Upon looking at that... It appears like the Digital ocean cert may not be using the correct flagging, as they're steps do not tell me to add the flag to the root CA, just to the `server-cert.pem` file. 

I'm going to try adding the flag to the CA and reimporting that into windows. Leaving these notes for posterity.

**Reply by tobiasbrunner** [reply]
> If I were to follow his steps exactly and keep names the same, the file I imported into windows would be `vpn.example.com.crt.pem`.

No, that's the server certificate. The CA certificate after those steps is `vpnca.crt.der`.

> Upon looking at that... It appears like the Digital ocean cert may not be using the correct flagging, as they're steps do not tell me to add the flag to the root CA, just to the `server-cert.pem` file.

That's correct. It should only be in the server cert (the CA certificate is not used to authenticate the server, only the server certificate).

**Reply by Brink2Three** [reply]
> No, that's the server certificate. The CA certificate after those steps is vpnca.crt.der.

Yep. You're correct. I misread that one. That's why my computer can't identify it, it's the wrong type of cert. 

> That's correct. It should only be in the server cert (the CA certificate is not used to authenticate the server, only the server certificate).

I hate to say it then, but that just fixed the problem. Adding `--flag serverAuth` to the ca generation from digital ocean fixed the issue. I can now connect without setting `DisableIKENameEkuCheck` to 1. 

The only step I changed from the Digital ocean guide was signing the root certificate, just by adding the flag as follows. 

```
pki --self --ca --lifetime 3650 --in ~/pki/private/ca-key.pem \
    --type rsa --dn "CN=VPN root CA" --flag serverAuth --outform pem > ~/pki/cacerts/ca-cert.pem
```

If I've misunderstood your instructions, please let me know, but since it is working on my test machine, I think that resolved the issue.

**Reply by tobiasbrunner** [reply]
That doesn't make sense. Unless you e.g. configured the CA certificate as server certificate in strongSwan. So I'm sorry to say but I'm not convinced this is an actual requirement and you didn't mess some of these steps up.

**Reply by Brink2Three** [reply]
> I'm not convinced this is an actual requirement and you didn't mess some of these steps up.

That is FULLY within the realm of possibility. If you want any additional information to confirm that, let me know and I can happily provide it. Though, Windows requires the CA cert to have that flag though, so that part of it makes sense. 

I regenerated the server cert and key using that CA, and now it works. I was able to connect windows 10 and 11 computers to strongswan and can route traffic through it.

**Reply by tobiasbrunner** [reply]
> If you want any additional information to confirm that, let me know and I can happily provide it.

If you could share the config (`swanctl.conf` and contents or `/etc/swanctl` subdirs with `pki --print` for the files in there, or if you use `ipsec.conf`, the same for that file and the contents of the `/etc/ipsec.d` subdirs), that would be appreciated. The exact Windows version(s) that didn't accept the certificate before would also be helpful (I think `winver` gives you that information).

> Though, Windows requires the CA cert to have that flag though, so that part of it makes sense.

I have to admit, it's not as clear as I thought. While [RFC 5280, section 4.2.1.12](https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.12) says about the extendedKeyUsage extension that `"In general, this extension will appear only in end entity certificates."`, the [CA/B Baseline Requirements](https://cabforum.org/baseline-requirements-documents/) (BR), which is what CAs and browsers use as guidelines for TLS certificates (and Microsoft always seemed to have treated IKE like TLS in regards to certificates), state that the extension MUST be encoded and contain serverAuth (and optionally clientAuth) in **intermediate/subordinate CA** certificates if the sub-CA issues end-entity/subscriber certificates (7.1.2.2.g). However, the BR also state that **root CA** certificates MUST NOT encode the extension (7.1.2.1.d). Now the question is, what if there is only a single level, that is, the root CA directly issues end-entity certificates. The document actually explicitly prohibits this e.g. in section 6.1.7, which definitely makes sense for large CAs. So I guess it's open to interpretation and, therefore, possible that some clients enforce that the extension is there for the CA certificate that issued the end-entity certificate, even if it's actually the root certificate (the BR don't actually specify how certificates are validated, only how they are to be issued).

It's just really weird that this wasn't an issue before. Looking at the history of the document, the specific wording in section 7.1.2.2.g actually changed with version 1.7.1 (from August 2020). Up until version 1.7.0 it stated: `"For Subordinate CA Certificates to be Technically constrained in line with section 7.1.5, then either the value serverAuth or clientAuth or both values MUST be present."` ("Technically Constrained" CA certificates with serverAuth also MUST include a [nameConstraints extension](https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.10) according to section 7.1.5.).

Since version 1.7.1, the document states: `"For all other Subordinate CA Certificates, including Technically Constrained Subordinate CA Certificates: This extension MUST be present ... For Subordinate CA Certificates that will be used to issue TLS certificates, the value serverAuth MUST be present."`. So the requirement was apparently extended to all intermediate CA certificates, not only those that are constrained. It's possible that Microsoft changed their trust chain validation in line with this, perhaps they previously only checked EKUs for CA certificates that contained a nameConstraints extension.

**Reply by Brink2Three** [reply]
EDIT: 
I realized I didn't include the base troubleshooting info: 
OS: Ubuntu 20.04 LTS
Kernel version (if applicable): 5.15.0-1029
strongSwan version(s): strongswan/focal-updates,focal-security,now 5.8.2-1ubuntu3.5 all
Tested/confirmed with the latest version: yes? (apt update does not list a new version other then the beta.) 

> If you could share the config (ipsec.conf and the contents of the /etc/ipsec.d subdirs), that would be appreciated.

Can do! This is gonna be a long one...
Here's ipconfig.conf: 

```
config setup
    charondebug="ike 1, knl 1, cfg 0"
    uniqueids=no

conn ikev2-vpn
    auto=add
    compress=no
    type=tunnel
    keyexchange=ikev2
    fragmentation=yes
    forceencaps=yes
    dpdaction=clear
    dpddelay=300s
    rekey=no
# Server side (Public facing)
    left=%any
    leftid=@vpnserver.example.net
    leftcert=server-cert.pem
    leftsendcert=always
    leftsubnet=0.0.0.0/0
# Client side (internal)
    right=%any
    rightid=%any
    rightauth=eap-mschapv2
    rightsourceip=10.100.100.2-10.100.100.254
    rightdns=10.10.10.10
    rightsendcert=never
    eap_identity=%identity
    ike=chacha20poly1305-sha512-curve25519-prfsha512,aes256gcm16-sha384-prfsha384-ecp384,aes256-sha1-modp1024,aes128-sha1-modp1024,3>
    esp=chacha20poly1305-sha512,aes256gcm16-ecp384,aes256-sha256,aes256-sha1,3des-sha1!
```

I don't know the standard practice for sharing PKI print details for active keys, so I just generated dummy certs using the same steps. 
The steps I used are as follows (These were copied directly to ipsec.d as shown in the last line.)

```
mkdir -p ~/pkitest/{cacerts,certs,private}

chmod 700 ~/pkitest

pki --gen --type rsa --size 4096 --outform pem > ~/pkitest/private/ca-key.pem

pki --self --ca --lifetime 3650 --in ~/pkitest/private/ca-key.pem \
    --type rsa --dn "CN=VPN root CA" --flag serverAuth --outform pem > ~/pkitest/cacerts/ca-cert.pem

pki --gen --type rsa --size 4096 --outform pem > ~/pkitest/private/server-key.pem

pki --pub --in ~/pkitest/private/server-key.pem --type rsa \
    | pki --issue --lifetime 1825 \
        --cacert ~/pkitest/cacerts/ca-cert.pem \
        --cakey ~/pkitest/private/ca-key.pem \
        --dn "CN=vpnserver.example.net" --san vpnserver.example.net \
        --flag serverAuth --outform pem \
    >  ~/pkitest/certs/server-cert.pem

sudo cp -r ~/pki/* /etc/ipsec.d/
```

Here is the outputs of pki print for the certs. I'm not sure how to output the keys, so I just did the CA and the server cert. 

```
pki --print --in ca-cert.pem
  subject:  "CN=VPN root CA"
  issuer:   "CN=VPN root CA"
  validity:  not before Mar 03 18:21:41 2023, ok
             not after  Feb 28 18:21:41 2033, ok (expires in 3649 days)
  serial:    05:76:23:1c:d1:32:74:55
  flags:     CA CRLSign serverAuth self-signed
  subjkeyId: b4:35:e0:af:8b:6c:15:02:3a:00:f2:5b:42:76:a7:03:c0:75:84:0d
  pubkey:    RSA 4096 bits
  keyid:     1d:e7:c2:22:9c:18:ae:68:2f:82:be:12:7b:ba:8c:eb:b8:83:34:70
  subjkey:   b4:35:e0:af:8b:6c:15:02:3a:00:f2:5b:42:76:a7:03:c0:75:84:0d

```
```
pki --print --in server-cert.pem                                                
  subject:  "CN=vpnserver.example.net"
  issuer:   "CN=VPN root CA"
  validity:  not before Mar 03 18:34:57 2023, ok
             not after  Mar 01 18:34:57 2028, ok (expires in 1824 days)
  serial:    35:0c:3b:5e:93:bb:79:5c
  altNames:  vpnserver.example.net
  flags:     serverAuth
  authkeyId: b4:35:e0:af:8b:6c:15:02:3a:00:f2:5b:42:76:a7:03:c0:75:84:0d
  subjkeyId: 64:94:36:7a:97:47:0b:e5:b4:ab:1f:ad:9b:d9:e6:84:e8:0e:0d:1e
  pubkey:    RSA 4096 bits
  keyid:     f7:78:9d:0e:75:64:e6:fb:50:24:37:0a:e1:27:91:85:23:c5:72:71
  subjkey:   64:94:36:7a:97:47:0b:e5:b4:ab:1f:ad:9b:d9:e6:84:e8:0e:0d:1e
```
>The exact Windows version(s) that didn't accept the certificate before would also be helpful (I think winver gives you that information)

My primary PC is running Windows 10 professional, Version 22H2. The Build ID is 19045.2604. 
My laptop is running Windows 11 Home, Version 22H2. The Build ID is 22621.1265
Both work perfectly with the new certs now. 

**LET ME KNOW** if you need anything else, I'll be happy to provide it. :)

**Reply by tobiasbrunner** [reply]
Thanks for the info. I tried using a VM with Windows 10 22H2 (same build) and wasn't able to reproduce the issue. A server certificate issued by a root CA certificate without serverAuth EKU was accepted just fine (both generated from scratch using the commands above, except for `--flag serverAuth` when generating the CA certificate).

Besides installing the CA certificate on the client, did you configure anything special e.g. via `Set-VpnConnection` or `Set-VpnConnectionIPsecConfiguration` PowerShell applets? Or via registry?

Where did you store the CA certificate? `Certificates (Local Computer)\Trusted Root Certification Authorities\Certificates`? When I do that and check the listing via `mmc`, I see "Intended Purposes" is set to `<All>`, which is what I would expect. With the old CA certificate, did you ever manually change its properties there (e.g. to "Disable all purposes for this certificate", or "Enable only the following purposes" with "Server Authentication" disabled)?

**Reply by Brink2Three** [reply]
> did you configure anything special e.g. via Set-VpnConnection or Set-VpnConnectionIPsecConfiguration PowerShell applets? Or via registry?

I scripted the certificate import and VPN profile creation via PowerShell. Here's the snippets related to those:

```
Import-Certificate `
    -CertStoreLocation cert:\LocalMachine\Root\ `
    -FilePath %USERPROFILE%\Documents\Strongswan\ca-cert.pem
```
This imports the cert into `Certificates (Local Computer)\Trusted Root Certification Authorities\Certificates`. 
```
Add-VpnConnection -Name "ExampleVPN" `
    -ServerAddress "vpnserver.example.net" `
    -TunnelType "IKEv2" `
    -AuthenticationMethod "EAP" `
    -EncryptionLevel "Maximum" `
    -RememberCredential `
	-DnsSuffix "internaldomain.com" `
```
```	
Set-VpnConnectionIPsecConfiguration -Name "ExampleVPN" `
    -AuthenticationTransformConstants GCMAES256 `
    -CipherTransformConstants GCMAES256 `
    -DHGroup ECP384 `
    -IntegrityCheckMethod SHA384 `
    -PfsGroup ECP384 `
    -EncryptionMethod GCMAES256
    -SplitTunneling $True
```
```
Add-VpnConnectionRoute -ConnectionName "CSE VPN to GCP" -DestinationPrefix 10.10.10.0/24
```

I wonder if the EncryptionLevel needs the CA cert role to match perfectly? 

> When I do that and check the listing via mmc, I see "Intended Purposes" is set to <All>, which is what I would expect. With the old CA certificate, did you ever manually change its properties there (e.g. to "Disable all purposes for this certificate", or "Enable only the following purposes" with "Server Authentication" disabled)?

Nope. I want to be able to deploy this with a Powershell script, so I'm intentionally not manually changing anything in MMC. One thing I did notice is that the cert I imported has a warning flag next to key usage in the details tab, and the intended purpose shows Server Authentication, not <All> under intended purposes. The EKU however does not have a warning flag, and has just the server auth OID. 

![image](https://user-images.githubusercontent.com/47770944/223168150-174a4726-c2ca-4737-b526-0c3d8b79bffc.png)

**Reply by tobiasbrunner** [reply]
> I scripted the certificate import and VPN profile creation via PowerShell.

I see. That seems to have the same effect as manually importing it (at least in my tests). "Intended Purposes" is still `<All>` and the authentication works fine.

> I wonder if the EncryptionLevel needs the CA cert role to match perfectly?

No idea what exactly this does or if it's relevant for IKEv2 connections (which can't really be negotiated without encryption and the cipher suites do not really seem to be affected, at least not in regards to DH). And if you set algorithms manually via `Set-VpnConnectionIPsecConfiguration` this will change from `Maximum` to `Custom` anyway (see `Get-VpnConnection`).

Don't know what exactly you mean with "matching the CA cert role". What role? Match in what way?

> One thing I did notice is that the cert I imported has a warning flag next to key usage in the details tab, and the intended purpose shows Server Authentication, not under intended purposes. The EKU however does not have a warning flag, and has just the server auth OID.

Yeah, that's the same here (also the case for many pre-installed CA certificates, by the way). Still works fine.

So I'm still not convinced anything changed here :)

**Reply by Brink2Three** [reply]
> Yeah, that's the same here (also the case for many pre-installed CA certificates, by the way). Still works fine.
So I'm still not convinced anything changed here :)

Again it's entirely possible that I'm doing something wrong. I'm new to strongswan and am much less familiar with the intricacies, but based on the fact that I had 2 other people try this, and only the flagged CA works, I tend to think it may be either Ubuntu/strongswan server related or certificate generation related. 

I'm wondering if the OID is different for serverAuth somehow...? If you try to add the code for serverAuth (1.3.6.1.5.5.7.3.1) manually, does it go through or say it's already present? 

When I try to add it to the cert that did not have the serverAuth flag during generation, it accepts the code and works. If I try to add that to the working cert, it fails, saying the EKU is already present.

**Reply by tobiasbrunner** [reply]
> I tend to think it may be either Ubuntu/strongswan server related or certificate generation related.

I used the commands you posted above to test this (just omitted the `--flag serverAuth` when generating the CA certificate). And I obviously used strongSwan as a server (I also don't see how the server could influence this, the certificate validation is a local matter). So I don't think that's it. Perhaps there was an issue with the CA or end-entity certificate you generated initially (did you use the same commands?). But I guess it's also possible that there is some, yet unknown, setting in Windows that causes this somehow.

> I'm wondering if the OID is different for serverAuth somehow...? If you try to add the code for serverAuth (1.3.6.1.5.5.7.3.1) manually, does it go through or say it's already present?

You mean via mmc? There it says "The OID you entered already exists" (which makes sense as "Server Authentication" is already in the list of purposes that can be selected). Or did you try via PowerShell? (If so, what command?)

> When I try to add it to the cert that did not have the serverAuth flag during generation, it accepts the code and works.

That's not the case here, at least not via mmc. I guess via PowerShell adding it just enables that particular purpose (same as selecting "Server Authentication" as purpose via mmc), maybe try checking the properties via mmc.

> If I try to add that to the working cert, it fails, saying the EKU is already present.

When I open the properties dialog for a CA certificate with serverAuth flag in mmc, I don't get the option to add new purposes. There is only "Sever Authentication" in the list of purposes, which could be disabled. (Compared to a long list of pre-defined purposes for the certificate without EKU.) This also makes sense as the note above the list explicitly states: "You may only edit certificate purposes that are allowed by the certification path."

**Comment by 5tefun** [comment]
I too have a problem connecting to a StrongSwan Ikev2 server. The fact is that I can connect on lan from android and windows 10, I can connect from internet on android, but i can't connect from internet using windows 11. Any ideea ? Both locations are behind NAT

**Comment by ip75** [comment]
I have exactly the same issue as @SaiPremaTech 
There are no logs on the win10 client. Only error in app event log. `The error code returned on failure is 13801.`

StrongSwan client for Android works like a charm.

This "solution" by Microsoft is not a solution at all.
https://learn.microsoft.com/en-us/troubleshoot/windows-server/networking/troubleshoot-always-on-vpn#error-13801-the-ike-authentication-credentials-are-invalid

rasclient does not answer anything to the strongswan server and logs a 13801 error in the event log.
```
2025.09.06 11:22:25 15[ENC1] <connection|1> generating IKE_AUTH response 1 [ IDr CERT AUTH EAP/REQ/ID ]
2025.09.06 11:22:25 15[NET1] <connection|1> sending packet: from xxx.xxx.xxx.xxx[4500] to 10.108.3.250[4500] (608 bytes)
2025.09.06 11:22:55 09[JOB1] <connection|1> deleting half open IKE_SA with 10.108.3.250 after timeout
2025.09.06 11:22:55 09[IKE2] <connection|1> IKE_SA bvgm[1] state change: CONNECTING => DESTROYING
```

https://docs.strongswan.org/docs/latest/interop/windowsCertRequirements.html#_disabling_extended_certificate_checks
`HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\RasMan\Parameters` - `DisableIKENameEkuCheck` = 1

0 | The EKU of the RRAS certificate is validated to check whether it contains the Server Auth (1.3.6.1.5.5.7.3.1) [RFC2459] Object Identifier (OID).
1 | The EKU of the RRAS certificate is not validated.

certificate has `serverAuth` flag but I get 13801 error. When `DisableIKENameEkuCheck` = 1 the error is gone. I guess this is a win10 RasMan bug.
